src/analysis/mesh/qgsmeshcontours.cpp

```diff
@@ -0,0 +1,384 @@
+         ( ( max_value < values[0] ) && ( max_value < values[1] ) && ( max_value < values[2] ) ) )
+      continue;
+
+    const bool valueInRange[3] =
+    {
+      ( min_value <= values[0] ) &&( max_value >= values[0] ),
+      ( min_value <= values[1] ) &&( max_value >= values[1] ),
+      ( min_value <= values[2] ) &&( max_value >= values[2] )
+    };
+
+
+    // all values are inside the range == take whole triangle
+    if ( valueInRange[0] && valueInRange[1] && valueInRange[2] )
+    {
+      QVector<QgsPointXY> ring = coords;
+      ring.push_back( coords[0] );
+      const QgsPolygonXY poly = {ring};
+      multiPolygon.push_back( QgsGeometry::fromPolygonXY( poly ) );
+      continue;
+    }
+
+    // go through all edges
+    QVector<QgsPointXY> ring;
+    for ( int i = 0; i < 3; ++i )
+    {
+      const int j = ( i + 1 ) % 3;
+
+      if ( valueInRange[i] )
+      {
+        if ( valueInRange[j] )
+        {
+          // whole edge is part of resulting contour polygon edge
+          if ( !ring.contains( coords[i] ) )
+            ring.push_back( coords[i] );
+          if ( !ring.contains( coords[j] ) )
+            ring.push_back( coords[j] );
+        }
+        else
+        {
+          // i is part or the resulting edge
+          if ( !ring.contains( coords[i] ) )
+            ring.push_back( coords[i] );
+          // we need to find the other point
+          double value = max_value;
+          if ( values[i] > values[j] )
+          {
+            value = min_value;
+          }
+          QgsPointXY xy = QgsGeometryUtils::interpolatePointOnLineByValue(
+                            coords[i].x(), coords[i].y(), values[i],
+                            coords[j].x(), coords[j].y(), values[j],
+                            value );
+          if ( !ring.contains( xy ) )
+            ring.push_back( xy );
+        }
+      }
+      else
+      {
+        if ( valueInRange[j] )
+        {
+          // we need to find the other point
+          double value = max_value;
+          if ( values[i] < values[j] )
+          {
+            value = min_value;
+          }
+          QgsPointXY xy = QgsGeometryUtils::interpolatePointOnLineByValue(
+                            coords[i].x(), coords[i].y(), values[i],
+                            coords[j].x(), coords[j].y(), values[j],
+                            value );
+          if ( !ring.contains( xy ) )
+            ring.push_back( xy );
+
+          // j is part
+          if ( !ring.contains( coords[j] ) )
+            ring.push_back( coords[j] );
+
+        }
+        else
+        {
+          // nothing here, we are outside the range
+          continue;
+        }
+      }
+    }
+
+    // add if the polygon is not degraded
+    if ( ring.size() > 2 )
+    {
+      ring.push_back( ring[0] );
+      const QgsPolygonXY poly = {ring};
+      multiPolygon.push_back( QgsGeometry::fromPolygonXY( poly ) );
```

Comment:
This method is very inefficient. See https://github.com/qgis/QGIS/commit/e70f11c4e1bbdc6b99165c35014a3112e95e0939 for the most efficient method for creating new polygon geometries.

Reply:
like the modified?